apps/manager_app/device_authentication.c

```diff
@@ -0,0 +1,367 @@
+      result->auth_device = sign_serial_number();
+      *result_valid = true;
+      next_state = SIGN_RANDOM_NUM;
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG:
+    case MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG:
+    default: {
+      usb_clear_event();
+      break;
+    }
+  }
+
+  return next_state;
+}
+
+static device_auth_state_e sign_random_handler(const manager_query_t *query,
+                                               manager_result_t *result,
+                                               bool *result_valid) {
+  pb_size_t request_type = get_request_type(&query->auth_device);
+  device_auth_state_e next_state = SIGN_RANDOM_NUM;
+
+  switch (request_type) {
+    case MANAGER_AUTH_DEVICE_REQUEST_INITIATE_TAG: {
+      usb_clear_event();
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG: {
+      uint8_t *challenge =
+          (uint8_t *)&(query->auth_device.challenge.challenge[0]);
+      result->auth_device = sign_random_challenge(challenge);
+      *result_valid = true;
+      next_state = RESULT;
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG: {
+      /* If SIGN_RANDOM_NUM == state, but the request_type received is
+       * MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG, it's an unexpected step in
+       * the flow. The device will treat it as an attempt to force device
+       * authentication status */
+      device_auth_handle_response(false);
+      result->auth_device = send_flow_complete();
+      *result_valid = true;
+      next_state = FLOW_COMPLETE;
+      break;
+    }
+    default: {
+      usb_clear_event();
+      break;
+    }
+  }
+
+  return next_state;
+}
+
+static device_auth_state_e result_handler(const manager_query_t *query,
+                                          manager_result_t *result,
+                                          bool *result_valid) {
+  pb_size_t request_type = get_request_type(&query->auth_device);
+  device_auth_state_e next_state = FLOW_COMPLETE;
+
+  switch (request_type) {
+    case MANAGER_AUTH_DEVICE_REQUEST_INITIATE_TAG:
+    case MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG: {
+      usb_clear_event();
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG: {
+      bool verified = query->auth_device.result.verified;
+      device_auth_handle_response(verified);
+      result->auth_device = send_flow_complete();
+      *result_valid = true;
+      next_state = FLOW_COMPLETE;
+      break;
+    }
+    default: {
+      usb_clear_event();
+      break;
+    }
+  }
+
+  return next_state;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void device_authentication_flow(const manager_query_t *query) {
+  if (false == check_manager_request(query, MANAGER_QUERY_AUTH_DEVICE_TAG)) {
+    // TODO: Handle bad data error
+    // TODO: Early flow exit
```

Comment:
For now atleast add a return here.

Reply:
and in the below condition

Reply:
I have added the handling as per our discussion today.